lib/type_check/internals/parser.ex

```diff
@@ -0,0 +1,268 @@
+      {Date, :diff, [1, 2]}
+      iex> alias Date, as: D
+      iex> ast_to_mfa(quote do: D.diff(1, 2))
+      {Date, :diff, [1, 2]}
+  """
+  @spec ast_to_mfa(Macro.t()) :: {module(), atom(), [Macro.t()]}
+  def ast_to_mfa({func, [context: _, import: module], args}) do
+    {module, func, args}
+  end
+
+  def ast_to_mfa({target, _, args}) do
+    case target do
+      {:., _, [{:__aliases__, [alias: false], [mod]}, func]} -> {elixir_module(mod), func, args}
+      {:., _, [{:__aliases__, [alias: mod], _}, func]} -> {mod, func, args}
+      {:., _, [{:__aliases__, _, [mod]}, func]} -> {elixir_module(mod), func, args}
+      {:., _, [mod, func]} when is_atom(mod) -> {mod, func, args}
+      _ -> raise("cannot infer function")
+    end
+  end
+
+  def ast_to_mfa(_), do: raise("not a function call")
+
+  defp elixir_module(module), do: String.to_atom("Elixir.#{module}")
+
+  @doc """
+  Convert the raw spec extracted by `fetch_spec/3` into type_check type.
+
+  The function is optimistic. If the type is not known, it assumes `any()`.
+
+      iex> import TypeCheck.Builtin
+      iex> import TypeCheck.Internals.Parser
+      iex> {:ok, spec} = fetch_spec(Kernel, :is_atom, 1)
+      iex> expected = function([term()], boolean())
+      iex> ^expected = convert(spec)
+  """
+  @spec convert(tuple()) :: TypeCheck.Type.t()
+  def convert(type), do: convert(type, %Context{default: B.any(), vars: %{}})
+
+  @spec convert(tuple(), Context.t()) :: TypeCheck.Type.t()
+
+  defp convert({:type, _, name, vars}, ctx), do: convert_type(name, vars, ctx)
+  defp convert({:atom, _, val}, _), do: B.literal(val)
+  defp convert({:integer, _, val}, _), do: B.literal(val)
+  defp convert({:ann_type, _, [_var, t]}, ctx), do: convert(t, ctx)
+  defp convert({:var, _, name}, ctx), do: Map.get(ctx.vars, name, ctx.default)
+
+  defp convert({:remote_type, _, [{:atom, _, module}, {:atom, _, type}, vars]}, ctx) do
+    case fetch_type(module, type, length(vars)) do
+      {:ok, spec, var_names} ->
+        # extract names of vars from raw spec
+        var_names = var_names |> Enum.map(fn {:var, _, name} -> name end)
+        # convert values from raw spec to type_check types
+        vars = vars |> Enum.map(&convert(&1, ctx))
+        # add new key-value pairs into existing vars
+        vars = Enum.zip(var_names, vars) |> Map.new()
+        vars = Map.merge(ctx.vars, vars)
+        convert(spec, %{ctx | vars: vars})
+
+      {:error, _} ->
+        ctx.default
+    end
+  end
+
+  defp convert(_, ctx), do: ctx.default
+
+  @spec convert_type(atom() | nil, list() | atom(), Context.t()) :: TypeCheck.Type.t()
+  # basic types
+  defp convert_type(:any, [], _), do: B.any()
+  defp convert_type(:atom, [], _), do: B.atom()
+  defp convert_type(:boolean, [], _), do: B.boolean()
+  defp convert_type(:pid, [], _), do: B.pid()
+  defp convert_type(:no_return, [], _), do: B.none()
+  defp convert_type(:none, [], _), do: B.none()
+
+  # unsupported by type_check yet
```

Comment:
Support for references and ports has been added, so this can be fixed :muscle: .